Postgres 13 release

lfittl opened the pull request:
Tracking PR for Postgres 13 support. Please note that this PR will not be merged, but instead the `13-latest` branch, as well as a tagged release, will be created when its completed.

## TODO

* [x] Query fingerprinting updates
    * [x] Use enum names in fingerprinting (instead of integer values)
    * [x] Review+test note for TransactionStmt.savepoint_name (https://github.com/lfittl/libpg_query/wiki/Fingerprinting)
    * [x] Review fingerprinting logic based on JSON format changes (due to Protobuf change)
    * [x] Copy over updated fingerprint tests + ensure C library works correctly
    * [x] Casts in WHERE clause https://github.com/pganalyze/pganalyze/issues/774
    * [x] Ignore digits in table names (\d{2,}) to group most time-partitioned tables together
    * [x] Sort BoolExpr.args, per https://github.com/lfittl/pg_query/issues/173
    * [x] Handle boolean values better ("SELECT is_deleted = $1" fingerprints differently than “SELECT is_deleted = false”)
    * [x] Ensure “SELECT * FROM users WHERE id = $1” and “SELECT * FROM users WHERE id IN ($1)” and “SELECT * FROM users WHERE id IN ($1, $2)” produce the same fingerprint (the first two get the same queryid in pg_stat_statements, but we currently record it as separate fingerprints -- similar for `NOT IN` and `!=`)
    * [x] Change list handling per https://github.com/lfittl/libpg_query/wiki/Fingerprinting#version-30-based-on-postgresql-13 (use hashes for de-dup/sorting instead of the full string value)
    * ~~[ ] Consider excluding default enum values from the fingerprint~~
        * ~~LIMIT_OPTION_DEFAULT~~
        * ~~SETOP_NONE~~
        * ~~DEFELEM_UNSPEC~~
        * ~~OVERRIDING_NOT_SET~~
        * ~~ONCOMMIT_NOOP~~
        * ~~NO_CHECK_OPTION~~
        * ~~SORTBY_NULLS_DEFAULT~~
        * ~~SORTBY_DEFAULT~~
        * ~~NO_CHECK_OPTION~~
        * ~~CTEMaterializeDefault~~
    * ~~[ ] Consider excluding TypeName.typemod = -1 (the default)~~
    * [x] Review whether any fingerprint format changes are needed to reduce the token copying in the C library for the non-debug fingerprinting case (right now any node-based field does a call to `_fingerprintInitForTokens` / `_fingerprintCopyTokens` in order to determine whether the field name should be printed -- we can likely avoid the copying with some form of backtracking, i.e. saving the prior state of the hash)
    * [x] Profile `test/complex` to see if there are any overheads that would require a fingerprint format change
    * [x] Consider mandating big endian output of the fingerprint hexadecimal format, to match xxhash's canonical form (see `XXH64_canonicalFromHash`)
    * [x] Consider using XXH3_64bits instead of XXH64, as its [more modern and faster](https://fastcompression.blogspot.com/2019/03/presenting-xxh3.html)
* [x] [Ruby library](https://github.com/lfittl/pg_query/pull/187)
    * [x] Convert Ruby library to new Protobufs
    * [x] Ruby Deparser
    * [x] Ruby Fingerprinting
    * [x] Linux test build for Ruby library
* [x] [Go library](https://github.com/lfittl/pg_query_go/pull/29)
    * [x] Convert Go library to new Protobuf-based structs
    * [x] Resolve Go library test failures
    * [x] Linux test build for Go library
* [x] Review musl build issue (#86)
* [x] Add missing asprintf include in `pg_query_fingerprint.c`
* [x] Review fingerprint function for collisions by comparing it with v2 fingerprints

## Release notes (work in progress)

Make sure to include the following:

* [x] Review notes in CHANGELOG here: https://github.com/lfittl/libpg_query/commit/e6268dc60009cd65fd6e812bb5970809c939b27f (these should still apply, except for default enum values not being output)
* [x] Examples of where field node name is no longer repeated:
    - ColumnDef.typeName
    - AlterRoleStmt.role
    - SelectStmt.withClause
    - AlterDefaultPrivilegesStmt.action (https://github.com/lfittl/libpg_query/issues/82)

lfittl committed:
Initial version of Postgres 12 support

Please note that this is still work in progress, and in particular the
fingerprinting logic will still change.

lfittl committed:
Add support for returning Postgres scanner result

This allows utilizing pg_query for use cases that need the raw token
information, instead of a parse tree. Due to additional modifications
to the Postgres source, this also contains information about comments
in the query string, and their location.

lfittl committed:
Extract source script: Rename blacklist => blocklist

lfittl committed:
Update to Postgres 13

lfittl committed:
Add support for ARM builds by explicitly copying ARM header file

Fixes #72

pkubaj committed:
Add arch-ppc.h for PPC architectures (#80)

lfittl committed:
Introduce new Protobuf parse tree output format

Up until now, this library relied on JSON to pass the parse result back
to the caller, which has a number of downsides, most importantly that
we don't have a readily available parser for JSON thats not tied to a
running Postgres server. That in turn makes it hard to provide
cross-language features such as deparsing directly in this library
(which would require reading back a parse tree that gets passed in).

Protobuf isn't perfect, but its straightforward enough to generate the
schema definitions for the parse tree nodes, and output the tree using
a bundled C protobuf library, which has a small enough SLOC count (~3k)
to not be noticeable in the big picture.

lfittl committed:
Scanner: Generate correct scan tokens based on yytokentype enum

Previously we used a hand-generated list that wasn't automatically
updated with new Postgres releases.

lfittl committed:
WIP for fingerprint changes to version 3

lfittl committed:
Protobuf output: Fix memory handling for strings

lfittl committed:
Extract headers: Correctly handle 3+ enum values in the same line

lfittl committed:
Fingerprinting: Update to v3 version (WIP)

lfittl committed:
Use single conds/defs file for both JSON and protobuf output, fix tests

lfittl committed:
Refresh parser/lexer patches for Postgres 13.1

This also fixes a few edge cases where parsing wasn't working as expected,
and resolves an issue that broke JSON operator parsing. The new set of
patches pass "make check" cleanly for Postgres, and also apply cleanly
to Postgres 13.1 to make pg_query-specific information work.

lfittl committed:
Normalize: Adjust logic for DefElem to work without parser patch

Previously we relied on a parser modification that would create A_Const
for DefElems in order to locate constant values for DefElem which have
a String argument.

In general, its good to reduce the number of parser patches, and this
particular case can be directly handled in the normalize method instead.

Note that the most prominent reason for doing this are password-related
utility commands, such as "ALTER USER test PASSWORD 'xyz'", which is
unfortunately not parameterized by pg_stat_statement's own logic.

lfittl committed:
Extract source: Correctly detect location of macOS header files

lfittl committed:
Fixup extract source script to be error-free, update ARM/PPC files

lfittl committed:
Merge protobuf files into one, introduce top-level parse result

lfittl committed:
Add experimental Protobuf C++ outfuncs, converge JSON output to match

This introduces the following changes to the JSON output format, in order
to match the Protobuf JSON mapping scheme:

- Default values are no longer output for Enum values
- Whitespace between control characters in JSON is no longer added
- "<" and ">" characters are escaped to avoid browser HTML injections

Separately, this commit also unifies all outfuncs related code into the
"pg_query_outfuncs" prefixed files, to make it easier to understand.

The Protobuf C++ build can be utilized by building with
"make USE_PROTOBUF_CPP=1". Note that the Protobuf C++ requirement is for
building only, and the library itself remains a C library (with a bundled
version of Protobuf-C).

lfittl committed:
Makefile: Add TEST_LDFLAGS to ensure libraries are listed last in build

lfittl committed:
Makefile: Don't specify -pedantic for C++ builds

lfittl committed:
Extract source with USE_ASSERT_CHECKING enabled

This ensures we have the necessary functions to compile an
assert-enabled build if necessary. Note that this doesn't mean that
asserts are enabled by default (they are not, you need to explicitly
use DEBUG=1).

lfittl committed:
Add Postgres patch for new MemoryContextDeleteTopLevel function

lfittl committed:
Add Postgres patch to make plpgsql_start/finish_datums available

This avoids a memory initialization problem that the PL/pgSQL code
otherwise has.

lfittl committed:
Free Postgres top-level memory context on thread exit / with function

Previously there was no way to free the top-level Postgres memory context,
causing threaded programs that churn through a lot of threads to leak
memory with each newly initialized thread-local top-level memory context.

Instead, this uses a newly introduced cleanup method to free the memory
when a pthread exits (note this causes a pthread dependency to be added
to this library). In addition, primarily for memory testing purposes, add a
new method "pg_query_exit" that performs the same cleanup on demand.

lfittl committed:
Makefile: Add VALGRIND=1 option, check protobuf version, other fixes

lfittl committed:
Fix memory leak issues and ensure code base has a clean Valgrind run

lfittl committed:
Switch to GitHub actions for tests, run Valgrind and Protobuf C++ tests

lfittl committed:
Enum defs: Set value field for all values by calculating it when needed

This makes it easier to map the integer back to the name, for example
when trying to understand which type a NodeTag value actually is.

lfittl committed:
GitHub Actions: Fix typos in flags handling

lfittl committed:
Update bundled protobuf-c to 1.3.3

lfittl committed:
Protobuf: Correctly set the oneof type for Node

If this isn't set, the oneof data isn't actually written out as expected.

lfittl committed:
Add automatically generated readfuncs to turn Protobuf into Postgres AST

This will be used in a subsequent commit to enable deparsing of the
Protobuf-encoded tree back into a SQL query.

lfittl committed:
JSON outfuncs: Fix handling of NIL list elements

Certain parse nodes intentionally fill a list with a NIL element, and
we didn't consistently output this the same way across Protobuf and
JSON implementations.

This is now output as an empty node, which should be considered a null
value. In protobuf this is represented as a Node message without a set
type (PG_QUERY__NODE__NODE__NOT_SET for Protobuf-C).

lfittl committed:
Add deparsing functionality that turns parse tree back into a SQL query

This is based on the deparser that was written over multiple years for
the pg_query Ruby library, and is now accessible for all bindings through
this new API and implementation.

Note that there are still many (mostly special case) queries that are
not yet supported by the deparser.

lfittl committed:
Copy Postgres regression test SQL files to be available in tests

lfittl committed:
Add new pg_query_split function to split up multi-statement strings

Naively one could assume that splitting a string by ";" is sufficient,
but it becomes tricky once one takes into consideration that this
character can also show up in identifier, constants or comments.

Therefore, use the Postgres parser to determine where each statement
starts, and how long it goes, within a given string.

lfittl committed:
Parse PL/pgSQL test: Close file descriptor in error cases

lfittl committed:
Improve deparser based on Postgres regression test files

This is a start to ensure our deparser works correctly by making use
of the SQL files provided in the Postgres source tree as part of the
regression test suite.

Note that for the regression tests we do not compare the output string,
because its not possible to return the exact same statement text due
to details such as whitespace and comments. However, we can compare
that parsing the resulting, deparsed query again yields an identical
parse tree (minus the location fields).

lfittl committed:
Separate split function into two: scanner and parser

These behave differently and can both be valuable, most importantly
when splitting statements in a file that may contain syntax errors that
cause a parser error, but are accepted by the scanner.

This particular situation is the case with the Postgres regression tests,
which is why we need this functionality for the deparser tests.

lfittl committed:
Always return Postgres version in result, even for empty queries

This is helpful for consistency, and makes it easy to pass the empty
query parse tree back to the deparser.

lfittl committed:
Extract headers: Correctly extract enums with char values

lfittl committed:
Correctly set non-standard Enum values when reading back protobuf

We previously assumed that all enums are zero-indexed and incrementing,
but that is not the case for some of the enum values, and this leads
to silent errors when reading back the parse trees.

Note that protobuf still ensures all enums are zero-indexed and
incrementing after that, because the protobuf format is best used that
way.

lfittl committed:
Readfuncs: Ensure CreateForeignTableStmt has the correct tag set

This node is a bit special because it is a superset of CreateStmt, and
thus the CreateStmt tag was being set by accident.

lfittl committed:
Deparser: Ensure complete regression test suite can be deparsed.

lfittl committed:
Fix split test memory allocation

lfittl committed:
Protobuf C++ outfuncs: Correctly translate enum values

lfittl committed:
Split with scanner: Correctly advance start position for empty statements

This matches the split with scanner behaviour to the split with parser
behaviour, which ensures that empty statements are ignored in the split.

lfittl committed:
Fix Valgrind run in GitHub Actions

lfittl committed:
Fix memory leaks in tests

lfittl committed:
Protobuf: Always start enum values at 1 instead of 0

This avoids the automatic skipping of zero-values in the JSON-mapped
version of the Protobuf output. To match, the simple JSON output
(without the Protobuf library), now also always outputs the enum value.

lfittl committed:
Improve output of Bitmapset fields, add to Protobufs

Note that Bitmapset fields are currently only used in parse analysis
trees, not raw parse trees, but this future-proofs our output code.

lfittl committed:
Update to Postgres 13.2 release

lfittl committed:
Use PgQueryProtobuf struct for scan result as well

lfittl committed:
PL/pgSQL: Output NEW/OLD variable numbers, record dno fields

Author: Ethan Resnick (https://github.com/ethanresnick)

lfittl committed:
Makefile: Allow passing in customized CFLAGS/PG_CONFIGURE_FLAGS/TEST_*

Author: Ethan Resnick (https://github.com/ethanresnick)

lfittl committed:
Examples: Use pg_query_exit to ensure clean Valgrind run

pyramation commented:
thanks @lfittl for taking this one! This release will be amazing and super useful for all the community 🚀

lfittl committed:
Improve parser patch that supports old ? pgss syntax

lfittl committed:
Move enum helper functions to dedicated file for easier inclusion

lfittl committed:
Fingerprinting: Output Enum values as strings

lfittl committed:
pg_query_parse: Always full structure with version field

This ensures empty results also contain the Postgres version number
directly in the parse tree.

lfittl committed:
Deparser: Error out instead of crash when target_list is incorrect

This would only occur when someone passes in a hand-constructed parse
tree.

lfittl committed:
Use XXH-based 64-bit integer for fingerprint, instead of SHA1-based string

This enables faster computation / better storage of the fingerprint, and
doesn't actually affect its usefulness much, because the typical
underlying data source for fingerprinted data (pg_stat_statements) also
uses a 64-bit integer to identify queries.

In addition, this vendors the xxhash library for computing the fingerprint,
and stops using the cryptographic SHA1 hash function for this, which is
unnecessarily slow. Note we don't use the Postgres hash function here,
because unlike pg_stat_statements we need the fingerprint to be portable
across different architectures.

For future proofing, all vendored libraries (protobuf-c and xxhash) are
now located in the new vendor/ directory. When not using the created
static library, make sure to include both protobuf-c and xxhash in the
build process for any libpg_query-based libraries.

lfittl committed:
Fingerprint: Ignore TypeCast when the arg is A_Const or ParamRef

This ensures a few edge cases are handled correctly, specifically
"SELECT a = $1" should have the same fingerprint as "SELECT a = false".

Whilst there might be value in considering typecasts as something that
is impactful enough to be included in the fingerprint, for now it seems
better to err on the side of too much grouping, and ignore all typecasts
if they contain a constant value or parameter reference.

lfittl committed:
Int to Enum helpers: Avoid returning invalid enum values (-1).

lfittl committed:
Fingerprinting: Treat A_Expr kind AEXPR_OP_ANY/AEXPR_IN as AEXPR_OP

This ensures that "x = $1", "x IN ($1)" and "x = ANY($1)" fingerprint
the same way.

lfittl committed:
Fingerprints: Use XXH3 64-bits hash instead of XXH64

Whilst the speed difference isn't currently noticable, there is no good
reason to use the slightly older XXH64 instead of the more modern XXH3
hash.

lfittl committed:
Fix Int To Enum default return values, attempt 2 :)

lfittl committed:
Fingerprinting: Ignore multiple subsequent digits in RangeVar.relname

This ensures that time-based partitioned tables, which usually include
the date in the table partition name, are treated as the same.

To avoid grouping together things like "table_v1" and "table_v2", we
require at least two subsequent digits.

This logic isn't perfect, but seems to cover the most common cases.

lfittl committed:
Fingerprinting: Optimize list and node field fingerprinting

For list fingerprints we were comparing the full token list of each node
alphabetically, which can be very expensive. Instead, run two passes of
fingerprinting, once to calculate the hash value into a sub context, and
then sort the list using that, running the fingerprint again with the
parent hash.

In addition, also update the rather inefficient "don't output node field
name if its value is empty" to use a simple backtracking mechanism that
stores the previous hash value.

These two optimizations combined bring the runtime of test/complex on a
2017 iMac Pro from ~7 seconds to ~4 seconds, likely due to the avoidance
of unnecessary string allocations and associated copying.

lfittl committed:
Fingerprint: Add explicit fingerprint_str result to avoid confusion

In case fingerprints are output, its preferable to output the canonical
form of the hash (XXH64_canonicalFromHash) represented as hex values.

This happens to be identical to printf("%016x", hash), but it seems less
error-prone to go directly through xxHash's canonical methods to protect
against any oversights in regards to endianness.

lfittl committed:
Fingerprint tests: Fix limitOptions => limitOption expected part name

lfittl committed:
Fingerprint: Don't output node name when node type is specified in parent

This matches the JSON format, which no longer presents the node names
when the parent node has a fixed type field, instead of the generic Node
type. This ultimately comes from the Protobuf format, which avoids an
unnecessary Node message when a field always has the exact same type
present for a field (the same way this is done in Postgres itself).

lfittl committed:
Fix fingerprint for test/complex test

lfittl committed:
Avoid dependency on execinfo library

lfittl committed:
Add missing stdio.h include in pg_query_fingerprint.c

lfittl committed:
Update README examples

lfittl committed:
Release 13-2.0.0

lfittl commented:
@pyramation @lelit This has been officially released now in `13-latest` - thanks for the early review!

Note that the repositories have also been moved into the pganalyze organization. No other changes to be expected, but this makes it clearer who funds this work going forward (my own company), and also reflects a more dedicated commitment to get new versions released more quickly going forward.

I'll also release a blog post tomorrow that summarizes the changes for this release, but I think you already know most of it anyway :)
